R/async.R

```diff
@@ -0,0 +1,131 @@
+
+runStepsIfForwarding <- function(initialValue, errorHandlerStep, steps) {
+  runStepsUntil(
+    initialValue = initialValue,
+    errorHandlerStep = errorHandlerStep,
+    conditionFn = function(x) {
+      !is_forward(x)
+    },
+    steps = steps
+  )
+}
+
+
+
+tryCatchWarn <- function(error, expr) {
+  oldWarn <- options("warn")[[1]]
+  tryCatch(
+    {
+      # Set to show warnings immediately as they happen.
+      options(warn=1)
+
+      force(expr)
+    },
+    error = error,
+    finally = {
+      options(warn = oldWarn)
+    }
+  )
+}
+
+
+runSteps <- function(initialValue, errorHandlerStep, steps) {
+  runStepsUntil(
+    initialValue,
+    errorHandlerStep,
+    function(x) {
+      # always continue
+      FALSE
+    },
+    steps
+  )
+}
+
+#' @import promises
+runStepsUntil <- function(initialValue, errorHandlerStep, conditionFn, steps) {
+  x <- initialValue
+
+  step_count <- length(steps)
+  nextStepPos <- 1L
+
+  runStep <- function() {
+    if (nextStepPos > step_count) {
+      return(x)
+    }
+
+    nextStep <- steps[[nextStepPos]]
+    nextStepPos <<- nextStepPos + 1L # TODO pass in as value? multisession issue
+    # if NULL is passed in (or not a function), it is skipped
+    if (!is.function(nextStep)) {
+      return(runStep())
```

Comment:
I'd do this for NULL, but throw if non NULL and not a function.

Reply:
fixed.